### ISSUE TITLE
SSL error on `doRead`

### ISSUE THREAD

**Issue by daniel-quintoandar** [http-kit/http-kit#469]
I'm getting `javax.crypto.BadPaddingException` exception which may be caused by [synchronization](https://bugs.openjdk.java.net/browse/JDK-8221218?focusedCommentId=14416634&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14416634). I'm not sure if this bug can be fixed on httpkit or just by updating java version :thinking: 

I'm using
- http-kit **2.5.3**
- java **openjdk version "11.0.11" 2021-04-20**

<details>
<summary>Exception stack</summary>

```
javax.crypto.BadPaddingException: Insufficient buffer remaining for AEAD cipher fragment (2). Needs to be more than or equal to IV size (8) + tag size (16)
  ?, in sun.security.ssl/decrypt
  ?, in sun.security.ssl/decodeInputRecord
  ?, in sun.security.ssl/decode
  ?, in sun.security.ssl/decode
  ?, in sun.security.ssl/decode
  ?, in sun.security.ssl/decode
  ?, in sun.security.ssl/readRecord
  ?, in sun.security.ssl/unwrap
  ?, in sun.security.ssl/unwrap
  ?, in javax.net.ssl/unwrap
  File "HttpsRequest.java", line 35, in org.httpkit.client/unwrapRead
    while ((res = engine.unwrap(peerNetData, peerAppData)).getStatus() == Status.OK) {
  File "HttpClient.java", line 191, in org.httpkit.client/doRead
    read = httpsReq.unwrapRead(buffer);
  File "HttpClient.java", line 494, in org.httpkit.client/run
    doRead(key, now);
  ?, in java.lang/run
```

</details>

**Comment by miikka** [comment]
We're seeing a `javax.net.SSLProtocolException` when using http-kit as a HTTP client. It's a different exception, but seems like it could be a related problem. It happens randomly in production and we do not have a reliable way to reproduce it in isolation.

The linked JDK issues look potentially relevant,. Unfortunately they're fixed in OpenJDK 11.0.12, which is going to be [released in July](https://wiki.openjdk.java.net/display/JDKUpdates/JDK11u#JDK11u-Timelines). Could be worth it to try an Early Access release, though.

* http-kit: 2.5.3 
* Java: Corretto-11.0.11.9.1 (build 11.0.11+9-LTS).

<details>
<summary>Exception stack</summary>
```clojure
{:error #error {
 :cause Input record too big: max = 16709 len = 20532
 :via
 [{:type javax.net.ssl.SSLProtocolException
   :message Input record too big: max = 16709 len = 20532
   :at [sun.security.ssl.SSLEngineImpl readRecord SSLEngineImpl.java 600]}]
 :trace
 [[sun.security.ssl.SSLEngineImpl readRecord SSLEngineImpl.java 600]
  [sun.security.ssl.SSLEngineImpl unwrap SSLEngineImpl.java 454]
  [sun.security.ssl.SSLEngineImpl unwrap SSLEngineImpl.java 433]
  [javax.net.ssl.SSLEngine unwrap SSLEngine.java 637]
  [org.httpkit.client.HttpsRequest unwrapRead HttpsRequest.java 35]
  [org.httpkit.client.HttpClient doRead HttpClient.java 191]
  [org.httpkit.client.HttpClient run HttpClient.java 494]
  [java.lang.Thread run Thread.java 829]]}}
```

</details>

**Comment by miikka** [comment]
We tried running Corretto-11.0.12.7.1, but the same problem keeps happening. Ah well.

The server we're connecting to is AWS CloudFront. I trust that their TLS implementation is not doing anything too wonky.

**Comment by borkdude** [comment]
Does anyone of you have a repro of the request? I could try with babashka which also has httpkit client but it compiled using GraalVM.

**Comment by miikka** [comment]
Unfortunately I don't have a reliable reproduction apart from running our code in our production environment. Some details here, though:

* The request is basic HTTPS GET request with some query parameters and a cookie header.
* A lot of these requests are being made, potentially at the same time. I suspect it's a concurrency issue somewhere.
* The requests are made to the same host using the same http-kit client instance. SNI is enabled with the code below:

```clojure
(require '[org.httpkit.client :as http])
(import '[java.net URI] '[javax.net.ssl SNIHostName SSLEngine SSLParameters])

(defn sni-enabled-ssl-configurer
  [^SSLEngine ssl-engine ^URI uri]
  (let [^SSLParameters ssl-params (.getSSLParameters ssl-engine)]
    (.setServerNames ssl-params [(SNIHostName. (.getHost uri))])
    (.setSSLParameters ssl-engine ssl-params)
    (.setUseClientMode ssl-engine true)))

(def my-client (http/make-client {:ssl-configurer sni-enabled-ssl-configurer}))
```

**Comment by borkdude** [comment]
@miikka the SNI client in bb is already enabled by default. Perhaps you could try to run the request in bb and see if it still happens?

**Comment by seancorfield** [comment]
We're also seeing this from time to time in production, without the SNI enabled configurer, without a specific client instance created, on a plain `post` that we dereference "immediately" before continuing on. So it's just reusing the default client singleton.

**Comment by xwang1498** [comment]
> We're seeing a `javax.net.SSLProtocolException` (message: `Input record too big: max = 16709 len = 20532`) when using http-kit as a HTTP client.

Hi @miikka , did you ever figure out what's going on here?

I think I have a partial explanation. A Cloudflare-hosted service we talk to sometimes returns a plain http response (e.g. `HTTP/1.1 400 Bad Request`), despite it being a TLS connection. The TLS stack interprets these bytes as a TLS record, where bytes 3 & 4 (the `P/` in `HTTP/1.1`) are the length, plus 5 bytes for the header. 0x502f + 5 = **20532 bytes**!

**Comment by seancorfield** [comment]
This reminded me I should have followed up on my report from September 2021: we ended up switching from `http-kit` to Hato because of this, after trying fresh client instances and also updating our JDK -- and, yes, CloudFlare was in the mix for us as well.

**Comment by miikka** [comment]
@xwang1498, I never figured it out but your explanation makes a lot of sense! We didn't have Cloudflare, but I can't rule out some other service returning plain HTTP requests in some cases. Personally I've moved on but @huima check this out, you might find it interesting.

**Comment by huima** [comment]
Hi @xwang1498, out of curiosity how did you come up with this explanation and how did you observe that in the systems? I haven't observed Cloudfront doing something like that, but then again I haven't tried to look or find something like that either.

**Comment by seancorfield** [comment]
@huima I can only point out that CloudFront and CloudFlare seem to have several differences in behavior.

**Comment by xwang1498** [comment]
@miikka @huima @seancorfield
I think I've finally figured out what's going on here, and it's a nasty (but fixable!) bug in http-kit.

### tl;dr
If re-using a kept-alive connection fails for some reason (e.g. the remote side closed it), http-kit will incorrectly use the old ssl engine when making a new connection.

### Longer explanation
With the assistance of Wireshark, I found this sequence of events:
- We successfully request many times over a long-running kept-alive connection.
- At some point, this connection is closed by the remote end.
- When trying to re-use this connection, we realize that it has closed, so we open a new connection. However, we don't start a new TLS handshake, but rather immediately send encrypted data (because we're incorrectly re-using the old ssl engine that has already handshaken) on the new connection.
- To the remote (Cloudflare) side, this encrypted data looks like garbage, because we didn't do a new TLS handshake. So they respond with `HTTP/1.1 400 Bad Request` and close this new connection.
- Our TLS stack interprets those bytes as a large TLS packet and the request fails with a red herring `SSLProtocolException: Input record too big: max = 16709 len = 20532` error (see [above](https://github.com/http-kit/http-kit/issues/469#issuecomment-1129100036) for why it's 20532).
- A subsequent request will open a brand new connection with a new ssl engine. So it will succeed and the system carries on.

I've created pull request #489 that tries to address this problem.

**Comment by ptaoussanis** [comment]
@xwang1498 That's terrific news, thanks so much for doing this investigation - and for the PR! I really appreciate it, and am sure many other users will too!

Likewise thanks to everyone else who patiently provided info on this long-standing issue!

**Comment by borkdude** [comment]
I second what Peter said, this is excellent!

**Comment by ptaoussanis** [comment]
https://github.com/http-kit/http-kit/releases/tag/v2.6.0-RC1 is now out, and includes @xwang1498's fix.

Closing the issue for now, but everyone please feel free to re-open if there's any remaining trouble.
